### PR TITLE
Add padding layer.

### PR DESCRIPTION
**Pull request by hzxie**
I created padding layer because TinyDNN hasn't support add padding in convolutional layers.
The padding layer is similar to `tf.pad` in [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/pad).

### CODE REVIEW

**Comment by hzxie** [comment]
Here's the code for testing.

```
#include <cstdlib>
#include <iostream>
#include <vector>

#include "tiny_dnn/tiny_dnn.h"

class TestNet : public tiny_dnn::network<tiny_dnn::sequential> {
 public:
  explicit TestNet(const std::string &name = "")
    : tiny_dnn::network<tiny_dnn::sequential>(name) {
    using pad     = tiny_dnn::layers::pad;
    *this << pad(5, 5, 2, 1, 2);
  }
};

int main() {
  TestNet nn;
  
  int init_value = 0;
  tiny_dnn::vec_t in(5 * 5 * 2);
  for ( auto itr = in.begin(); itr != in.end(); ++ itr ) {
    *itr = init_value ++;
  }

  int cnt = 0;
  auto res = nn.predict(in);
  for ( auto itr = res.begin(); itr != res.end(); ++ itr ) {
    std::cout << *itr << "\t";
    if ( ++ cnt % 7 == 0 ) {
      std::cout << std::endl;
    }
    if ( cnt % 63 == 0 ) {
      std::cout << std::endl;
    }
  }
  std::cout << std::endl;
}
```

The output is listed as follows

```
0	0	0	0	0	0	0	
0	0	0	0	0	0	0	
0	0	1	2	3	4	0	
0	5	6	7	8	9	0	
0	10	11	12	13	14	0	
0	15	16	17	18	19	0	
0	20	21	22	23	24	0	
0	0	0	0	0	0	0	
0	0	0	0	0	0	0	

0	0	0	0	0	0	0	
0	0	0	0	0	0	0	
0	25	26	27	28	29	0	
0	30	31	32	33	34	0	
0	35	36	37	38	39	0	
0	40	41	42	43	44	0	
0	45	46	47	48	49	0	
0	0	0	0	0	0	0	
0	0	0	0	0	0	0
```

**Comment by beru** [comment]
@hzxie 

> Here's the code for testing.

Adding unit tests would be better. In that way, future changes can be tested.

travis-ci jobs failed.
https://travis-ci.org/tiny-dnn/tiny-dnn/builds/417261964?utm_source=github_status&utm_medium=notification

Running clang-format with `scripts/clang-format.sh` is required. But as I mentioned in https://github.com/tiny-dnn/tiny-dnn/pull/978#issuecomment-414088481 , this requirement is cumbersome and I personally think this task should be removed for contributors.

**Comment by hzxie** [comment]
@beru @edgarriba 
All unit tests for padding layer is created and passed.
BTW, CI often failed due to the failure of unit testing cases named `train2`.

Maybe we should make a larger threshold, 1e-4 is too small?
See: https://travis-ci.org/tiny-dnn/tiny-dnn/builds/419060831
screenshot from 2018-08-22 17-02-02

**Comment by hzxie** [comment]
@edgarriba @beru 
Please come here to merge my PR.
Thank you.

**Comment by edgarriba** [comment]
@hzxie can you investigate this issue with train2 ? I remember @prlz77 mentioned at some point that this was due to the gradient check he implemented.

**Comment by hzxie** [comment]
Actually, #978 also failed on train2 sometimes.
In this PR, there’s nothing related to train2.
I think I can fix this issue in another PR.

**Comment by hzxie** [comment]
@edgarriba
Can you give me some hints to fix this problem? Because train2 only failed in some platforms.
What did @prlz77 said?

**Comment by hzxie** [comment]
@edgarriba 
It seems that CI errored on all platforms.

Here's the error log:
```
CMake Error in CMakeLists.txt:
  export called with target "gmock" which requires target "gtest" that is not
  in the export set.
  If the required target is not easy to reference in this call, consider
  using the APPEND option with multiple separate calls.
-- Generating done
-- Build files have been written to: /home/travis/build/tiny-dnn/tiny-dnn
The command "if [ "$TRAVIS_OS_NAME" == "linux" ] && [ "$CXX" == "g++-4.9" ]; then lcov --directory . --zerocounters; cmake -DUSE_TBB=$USE_TBB -DUSE_SSE=$USE_SSE -DUSE_AVX=$USE_AVX -DUSE_DOUBLE=$USE_DOUBLE -DBUILD_TESTS=$BUILD_TESTS -DCOVERALLS=$COVERALLS -DUSE_ASAN=ON -DBUILD_EXAMPLES=$BUILD_EXAMPLES .; fi" failed and exited with 1 during .
Your build has been stopped.
```

**Comment by hzxie** [comment]
#986 also failed.
Maybe there's something wrong with TravisCI?

**Comment by beru** [comment]
@hzxie It looks like recent googletest change is the cause. https://github.com/jaagr/polybar/issues/1393

I'm wondering if adding
```
option(INSTALL_GMOCK "Install Googletest's GMock?" OFF)
option(INSTALL_GTEST "Install Googletest's GTest?" OFF)
```
before https://github.com/tiny-dnn/tiny-dnn/blob/ceb8b9f04469bb03423c55a67bc164c3a52908d1/test/CMakeLists.txt#L14
might solve the problem...

**Comment by beru** [comment]
Talking about unit tests.

https://github.com/catchorg/Catch2 is also famous test framework. But I guess rewriting unit tests isn't trivial.

**Comment by edgarriba** [comment]
@beru Catch2 looks pretty good and fits perfect with our header-only policy. If anyone wants to take this task, why not.

**Comment by hzxie** [comment]
Rewriting unit testing cases is time-consuming.
I think we need to fix CI failure first.

**Comment by edgarriba** [comment]
@hzxie agree. This PR https://github.com/tiny-dnn/tiny-dnn/pull/968 tries to fix this, take a look at it. Besides, what makes sense now is to rethink about how to setup a proper infrastructure with travis since lately things became overcomplicated with the more third-party stuff we added. At some point we should introduce docker based testing. We internally had this discussion. We should probably start a proper discussion about it. Let me open a thread since there are more some folks interested in this as well @Ravenwater @chrberger

**Comment by beru** [comment]
@hzxie 

> Rewriting unit testing cases is time-consuming.
I think we need to fix CI failure first.

I initially thought so but couldn't figure out the cause of googletest error with cmake.
So I've decided to change test framework to Catch2.

Please review #987.

**Comment by beru** [comment]
> It seems that CI errored on all platforms.

Actually, it's OK with VC++ (Windows platform) build but not sure why.

**Comment by beru** [comment]
@edgarriba 
> Catch2 looks pretty good and fits perfect with our header-only policy. If anyone wants to take this task, why not.

I want to remove CI job problems, so made a pull request. #987

> UPDATE: I suggest discussing it first in
https://groups.google.com/forum/#!forum/tiny-dnn-dev

Maybe next time.

**Comment by hzxie** [comment]
@beru 

The AUTHORS file has been updated.
I will push it after PR #987 is merged.

BTW, I suggest that the last line in this file should be changed to:

```
Generated by https://github.com/nodejs/node/blob/master/tools/update-authors.sh
```

What do you think?

**Comment by edgarriba** [comment]
@hzxie why don't we just reuse this script ?

**Comment by beru** [comment]
@hzxie 

> BTW, I suggest that the last line in this file should be changed to:
>```
> Generated by https://github.com/nodejs/node/blob/master/tools/update-authors.sh
>```
> What do you think?

I think it's fine. In that way, we can show homage to nodejs team.

It's a short script file, so adding the file to `scripts` folder seems OK.
In that way, we'll be able to customize the file. And we can also mention where it originally came from in a commit message.

**Comment by beru** [comment]
@edgarriba 
> why don't we just reuse this script ?

The script appends comments after authors list but those comments were manually removed.

https://github.com/tiny-dnn/tiny-dnn/commits/master/AUTHORS

It would be easier to understand if you run the script.

**Comment by edgarriba** [comment]
@beru reuse or adapt

**Comment by beru** [comment]
@edgarriba 
> reuse or adapt

I see. So adding the file to scripts folder seems fine.
We can leave tail comments in a natural way. It can be adjusted as well.

**Comment by hzxie** [comment]
@beru @edgarriba 
Please review this PR.
I updated my test code and added the script for updating authors.

BTW, the new test script is not clear as before.
I don't know whether my newly added test script is run or not.
I think we should print the name of each test case before we run each test case.

**Comment by edgarriba** [comment]
@hzxie  LGTM. However, you should split the commit one for padding and the other to update the authors script.

**Comment by hzxie** [comment]
@edgarriba 
The script for updating authors was removed.
The AUTHORS file was reverted to the original version.

Please consider merging this PR.

**Comment by hzxie** [comment]
@edgarriba @beru 
Also please review PR #989.

File: .gitignore [hunk 1]
```diff
@@ -219,4 +219,4 @@ docs/_build
 *.pb.h
 CMakeFiles*
 build*
-DartConfiguration.tcl
+DartConfiguration.tcl
```
Comment (beru) [inline .gitignore]: No new line at end of file

File: AUTHORS [hunk 2]
```diff
@@ -72,4 +72,4 @@ Karan Desai <karandesai_96@live.com>
 MeiHui FAN <mhfan@ustc.edu>
 Rui Huang <vowstar@users.noreply.github.com>
 Siddhartha Rao Kamalakara <srk97c@gmail.com>
-Jaakko Rantala <jaakko.rantala@gmail.com>
+Jaakko Rantala <jaakko.rantala@gmail.com>
```
Comment (beru) [inline AUTHORS]: No new line at end of file

File: tiny_dnn/tiny_dnn.h [hunk 3]
```diff
@@ -102,6 +103,8 @@ using fc = tiny_dnn::fully_connected_layer;
 
 using dense = tiny_dnn::fully_connected_layer;
 
+using zreo_pad = tiny_dnn::zero_pad_layer;
```
Comment (beru) [inline tiny_dnn/tiny_dnn.h]: I've found  a typo. `zreo_pad` should be `zero_pad`.

Comment (hzxie) [reply tiny_dnn/tiny_dnn.h]: OMG.
Thank you!

**Pull request merged**